### PR TITLE
migrates to the new contact-info

### PR DESCRIPTION
**Pull request by behzadnouri**
#### Problem
move on to new contact-info

#### Summary of Changes
replaced uses of legacy contact-info with the new one.

### CODE REVIEW

File: local-cluster/tests/local_cluster.rs [hunk 1]
```diff
@@ -5706,6 +5703,7 @@ fn test_randomly_mixed_block_verification_methods_between_bootstrap_and_not() {
 
 /// Forks previous marked invalid should be marked as such in fork choice on restart
 #[test]
+#[ignore]
 #[serial]
 fn test_invalid_forks_persisted_on_restart() {
```
Comment (behzadnouri) [inline local-cluster/tests/local_cluster.rs:5708]: @AshwinSekar this test fails with this commit.
It passes if I make below change to push contact-infos out more frequently.
```diff
diff --git a/gossip/src/cluster_info.rs b/gossip/src/cluster_info.rs
index 2707a3ea0b..02aa95c1ea 100644
--- a/gossip/src/cluster_info.rs
+++ b/gossip/src/cluster_info.rs
@@ -1875,7 +1875,7 @@ impl ClusterInfo {
                     entrypoints_processed = entrypoints_processed || self.process_entrypoints();
                     //TODO: possibly tune this parameter
                     //we saw a deadlock passing an self.read().unwrap().timeout into sleep
-                    if start - last_push > CRDS_GOSSIP_PULL_CRDS_TIMEOUT_MS / 2 {
+                    {
                         self.push_self();
                         self.refresh_push_active_set(
                             &recycler,
```

Another way to make it pass is to use the new contact-infos in pull-requests as below, which would also help to propagate contact-infos (but we cannot do this yet because of compatibility issues).
```diff
diff --git a/gossip/src/cluster_info.rs b/gossip/src/cluster_info.rs
index 2707a3ea0b..c59a8713cb 100644
--- a/gossip/src/cluster_info.rs
+++ b/gossip/src/cluster_info.rs
@@ -1564,11 +1564,7 @@ impl ClusterInfo {
             .map(|(_, filters)| filters.len())
             .sum::<usize>() as u64;
         self.stats.new_pull_requests_count.add_relaxed(num_requests);
-        // TODO: Use new ContactInfo once the cluster has upgraded to:
-        // https://github.com/anza-xyz/agave/pull/803
-        let self_info = LegacyContactInfo::try_from(&self.my_contact_info())
-            .map(CrdsData::LegacyContactInfo)
-            .expect("Operator must spin up node with valid contact-info");
+        let self_info = CrdsData::ContactInfo(self.my_contact_info());
         let self_info = CrdsValue::new_signed(self_info, &self.keypair());
         let pulls = pulls
             .into_iter()
```


So I am guessing the test relies on contact-infos being propagated by a certain time or something. I have ignored it for now not to block this change because that is technically not a problem with this code. Can you please take a look?

Comment (AshwinSekar) [reply local-cluster/tests/local_cluster.rs:5708]: Will do, thanks for looking.

File: gossip/src/crds_value.rs [hunk 2]
```diff
@@ -155,7 +155,7 @@ impl CrdsData {
         // TODO: Assign ranges to each arm proportional to their frequency in
         // the mainnet crds table.
         match kind {
-            0 => CrdsData::LegacyContactInfo(LegacyContactInfo::new_rand(rng, pubkey)),
+            0 => CrdsData::ContactInfo(ContactInfo::new_rand(rng, pubkey)),
```
Comment (gregcusack) [inline gossip/src/crds_value.rs:158]: since we haven't removed `LegacyContactInfo` completely from the network, should we make `ContactInfo` an additional field rather than in place of `LegacyContactInfo`? will help ensure backwards compatability until `LegacyContactInfo` is fully gone

File: gossip/src/crds_entry.rs [hunk 3]
```diff
@@ -52,7 +52,7 @@ impl_crds_entry!(CrdsValue, |entry| Some(&entry?.value));
 impl_crds_entry!(VersionedCrdsValue, |entry| entry);
 
 // Lookup by Pubkey.
-impl_crds_entry!(LegacyContactInfo, CrdsData::LegacyContactInfo(node), node);
+impl_crds_entry!(ContactInfo, CrdsData::ContactInfo(node), node);
```
Comment (gregcusack) [inline gossip/src/crds_entry.rs:55]: are we going to run into an issue where we cannot lookup `LegacyContactInfo` from the `crds.table`?

File: gossip/src/cluster_info.rs [hunk 4]
```diff
@@ -1571,6 +1564,8 @@ impl ClusterInfo {
             .map(|(_, filters)| filters.len())
             .sum::<usize>() as u64;
         self.stats.new_pull_requests_count.add_relaxed(num_requests);
+        // TODO: Use new ContactInfo once the cluster has upgraded to:
+        // https://github.com/anza-xyz/agave/pull/803
         let self_info = LegacyContactInfo::try_from(&self.my_contact_info())
             .map(CrdsData::LegacyContactInfo)
             .expect("Operator must spin up node with valid contact-info");
```
Comment (gregcusack) [inline gossip/src/cluster_info.rs:1571]: i do not understand why this is incompatable. looks like we're able to handle the ContactInfo on the receive side of a `PullRequest` in `handle_batch_pull_requests()`: https://github.com/behzadnouri/solana/blob/d6d95f92fbd311efabf51514062427d4dcb479fc/gossip/src/cluster_info.rs#L1967

File: local-cluster/tests/local_cluster.rs [hunk 5]
```diff
@@ -5706,6 +5703,7 @@ fn test_randomly_mixed_block_verification_methods_between_bootstrap_and_not() {
 
 /// Forks previous marked invalid should be marked as such in fork choice on restart
 #[test]
+#[ignore]
 #[serial]
```
Comment (gregcusack) [inline local-cluster/tests/local_cluster.rs:5707]: scares me a little that this test is failing just from switching to `ContactInfo`. Makes me think it's not getting handled properly somewhere...

**Review by gregcusack — COMMENTED** [review]
just a few questions/comments. Also, do we still want to boost `LegacyContactInfo` when generating Pull Responses? https://github.com/behzadnouri/solana/blob/d6d95f92fbd311efabf51514062427d4dcb479fc/gossip/src/cluster_info.rs#L2124

File: gossip/src/cluster_info.rs (hunk 4 above)
Comment (behzadnouri) [reply gossip/src/cluster_info.rs:1571]: Right, but that was only done in https://github.com/anza-xyz/agave/pull/803 which is not backported to v1.17 yet. So mainnet node still expect legacy contact-info in the pull request and otherwise drop the pull request.
To keep master branch compatible with mainnet, we cannot yet switch to the new contact-info here until mainnet has upgraded to #803 .

File: gossip/src/crds_entry.rs (hunk 3 above)
Comment (behzadnouri) [reply gossip/src/crds_entry.rs:55]: If there were any instances of that, we would get compile-time error.

File: gossip/src/crds_value.rs (hunk 2 above)
Comment (behzadnouri) [reply gossip/src/crds_value.rs:158]: I think it should be fine.
This commit is reducing the uses of LegacyContactInfo to pull-requests.
We are not going to backport this code, so by the time this code is about to hit mainnet, the cluster has already upgraded to https://github.com/anza-xyz/agave/pull/803 and we can remove legacy contact-info from pull-requests as well.

File: local-cluster/tests/local_cluster.rs (hunk 5 above)
Comment (behzadnouri) [reply local-cluster/tests/local_cluster.rs:5707]: yeah, but looks like if I make either of patches mentioned in https://github.com/anza-xyz/agave/pull/823/files#r1575382916 it passes. so looks like it relies on contact-infos propagation by a certain time which is not guaranteed anyways.

Nonetheless, we are not going to backport this patch, and as soon as mainnet has upgraded to https://github.com/anza-xyz/agave/pull/803 we can use the new contact-info in the pull requests which makes the test pass.

But agreed that we should investigate this test and figure out what assumptions are failing. Ashwin has opened https://github.com/anza-xyz/agave/issues/982 to keep track of this.

File: gossip/src/cluster_info.rs (hunk 4 above)
Comment (gregcusack) [reply gossip/src/cluster_info.rs:1571]: ahh good to know we need to keep master compatible with mainnet. sounds good

File: local-cluster/tests/local_cluster.rs (hunk 5 above)
Comment (gregcusack) [reply local-cluster/tests/local_cluster.rs:5707]: ya good point. the frequency fix is interesting. Wondering why that affects the ContactInfo but not LegacyContactInfo? I'd think it wouldn't have an affect...

File: gossip/src/cluster_info.rs (hunk 4 above)
Comment (gregcusack) [reply gossip/src/cluster_info.rs:1571]: is this the same reason we keep adding `LegacyContactInfo` to PullResponses with higher probability: https://github.com/behzadnouri/solana/blob/d6d95f92fbd311efabf51514062427d4dcb479fc/gossip/src/cluster_info.rs#L2124?

**Review by gregcusack — APPROVED** [review]
lgtm. just need to keep an eye on https://github.com/anza-xyz/agave/issues/982

File: local-cluster/tests/local_cluster.rs (hunk 5 above)
Comment (behzadnouri) [reply local-cluster/tests/local_cluster.rs:5707]: > the frequency fix is interesting. Wondering why that affects the ContactInfo but not LegacyContactInfo? 

Previously, in addition to push and pull-responses, one more way that contact-infos were propagated across cluster was through the "caller" in pull-requests:
https://github.com/anza-xyz/agave/blob/e7452b768/gossip/src/cluster_info.rs#L2076-L2081
https://github.com/anza-xyz/agave/blob/e7452b768/gossip/src/crds_gossip_pull.rs#L296-L307
`process_pull_requests` is terribly named but basically what it does it upserts CRDS table with the caller's contact-info attached in pull-requests.

With this commit we are changing what struct counts as "contact-info" by switching from `LegacyContactInfo` to `ContactInfo`. So that caller's (legacy) contact-info attached to pull-requests is no longer relevant (because it is not a `ContactInfo`, but a `Legacy...`).

So basically with this commit we are killing that extra propagation path for contact-infos I mentioned earlier above. Increasing the frequency of "push_self" seems to compensate for that.

File: gossip/src/cluster_info.rs (hunk 4 above)
Comment (behzadnouri) [reply gossip/src/cluster_info.rs:1571]: yeah, since the cluster is still relying on `LegayContactInfo` we need to keep them prioritized for now.

File: local-cluster/tests/local_cluster.rs (hunk 5 above)
Comment (gregcusack) [reply local-cluster/tests/local_cluster.rs:5707]: ahhhh yes got it. that makes sense
